### PR TITLE
feat(doc-gates): enforce docs structure, frontmatter, and generated index

### PR DESCRIPTION
**Pull request by DeJeune**
> ### Branch strategy
>
> - Active development targets `main`.

**Stacked on #18856** (← #18848 ← #18844 ← #18843) — the gates layer of the docs-governance proposal; review the stack bottom-up.

### What this PR does

Before this PR: nothing verifies that docs describe code that still exists (the rot class fixed in #18844 was caught by hand), the `docs/README.md` index is hand-maintained and drifts, and CI runs no doc check at all.

After this PR:

- Every doc under `docs/references/**` carries frontmatter: a one-line `description` and `sources` (the code paths it describes); `docs/contrib/**` needs only `description`. ~110 docs annotated; every `sources` path verified to exist.
- Three gates (tsx scripts with exported functions + Vitest coverage): `verify-doc-structure` (closed domain set at the references root, every domain owns a `README.md` — seven domains gain homes: chat, components, file, knowledge, testing, security, provider-model), `verify-doc-frontmatter` (required fields; **a `sources` path that stopped existing fails the build** — silent rot becomes a red check), and `gen-doc-index --check` (`docs/README.md` is now generated from frontmatter; hand-edits fail).
- `pnpm docs:check` aggregates links + the three gates, joins `ci:basic-check` (closing the CI gap) and replaces the bare link check in `build:check`. `CLAUDE.md` documents the workflow (`pnpm docs:index` after frontmatter edits).

Fixes: N/A

### Why we need it and why it was done in this way

`sources` also enables reverse lookup — diff paths ∩ sources = "docs this PR should update" — planned for `gh-pr-review` in the proposal's Phase 4. Frontmatter admission rule (recorded in #18843): only fields the path, H1, or git cannot carry, with a named script consumer; `domain`/`title`/`updated`/`status` fields were rejected on that rule.

The following tradeoffs were made:

- `sources` prefers directories over files (fewer false breaks on refactors); precision is per-doc judgment.
- The generated index takes titles from each doc's H1 verbatim (including two emoji headings) — H1 cleanup belongs to the content-audit phase, not the generator.
- Annotation surfaced stale body citations (chat adapters target-architecture dirs, command docs citing `src/shared/command/`, an ai transport rename); frontmatter uses current paths, body fixes are deferred to the per-domain content audits.

The following alternatives were considered: encoding metadata in paths alone, dsh-style (rejected in #18843 — the rot gate and generated index both need per-file machine-readable fields); a hardcoded domain list in the gate vs deriving from disk (kept the explicit list — that is what makes the set *closed*).

Links to places where the discussion took place: #18843

### Breaking changes

None for users. For contributors: `docs/README.md` is now generated (edit frontmatter + `pnpm docs:index`), and new reference docs must carry frontmatter or CI fails.

### Special notes for your reviewer

Verification: new gate tests 10/10 (`pnpm test scripts/__tests__/verify-doc-structure.test.ts scripts/__tests__/verify-doc-frontmatter.test.ts scripts/__tests__/gen-doc-index.test.ts`); `pnpm docs:check` fully green; `pnpm lint` green (typecheck node/web/aicore, i18n, biome). Frontmatter blocks are pure insertions — every doc body is byte-identical (diff shows 0 deletions on annotated files).

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Branch: This PR targets `main`
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
NONE
```

### CODE REVIEW

**Review by vaayne — APPROVED** [review]

**Comment by kangfenmao** [comment]
@claude review

**Comment by DeJeune** [comment]
Fixed a real gap found by review on #18843: **this PR did not actually put the docs gate in CI.**

`.github/workflows/ci.yml` never invokes `pnpm ci:basic-check` — its `basic-checks` job inlines the individual checks through `concurrently`, so adding `docs:check` to that script alone would have run the gate nowhere. Added it as a fourth parallel lane in the `Read-only Checks` step (with its `--names` / `--prefix-colors` entries), so it now runs on every pull request. The `ci:basic-check` script keeps the entry so the local equivalent stays honest.

Verified: workflow parses (`yaml.parse`), and `pnpm docs:check` is green.

**Comment by DeJeune** [comment]
Fixed, though the diagnosis needs one correction — the two scenarios do not both hold.

**Scenario 2 (`..` traversal) is real.** `path.join(repoRoot, "../../secret.txt")` does resolve outside the repo, and if that path happens to exist the entry passes the existence check. Fixed.

**Scenario 1 (absolute path) does not reproduce.** An absolute second argument winning is `path.resolve` semantics, not `path.join`:

```
path.join("/repo", "/etc/passwd")    → /repo/etc/passwd
path.resolve("/repo", "/etc/passwd") → /etc/passwd
```

So `/etc/passwd` was already rejected today — it becomes `/repo/etc/passwd`, which does not exist. The proposed fix switching to `path.resolve` would have *introduced* that behavior and then caught it with the `startsWith` guard; same end state, longer route.

I also would not frame this as a security issue. `sources` is committed frontmatter in a repo where anyone who can edit it can already edit the gate script itself, and the gate only calls `existsSync`. The defect worth fixing is a **contract** one: diff paths are repo-relative, so an absolute or escaping entry can never match one — it is meaningless to the doc and silently useless to the Phase 4 reverse lookup.

Fix (`d0ac95c`): reject both forms by **format**, before the existence check, so they get an accurate message instead of a misleading "does not exist":

```ts
// Diff paths are repo-relative, so an absolute or escaping entry can never match one.
if (path.isAbsolute(source) || source.split(/[/\\]/).includes("..")) {
  failures.push(`${rel}: \`sources\` entry is not repo-relative: ${source}`)
} else if (!fs.existsSync(path.join(repoRoot, source))) {
  ...
}
```

Two tests added. The traversal one points at a file that **does** exist outside the repo root — otherwise it would pass for the wrong reason and pin nothing. Verified: `scripts` project green (22 files / 226 tests), `pnpm docs:check` green.

**Review by kangfenmao — APPROVED** [review]

File: scripts/verify-doc-frontmatter.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  } else if (description.includes('\n')) {
+    failures.push(`${rel}: \`description\` must be a single line`)
+  }
+
+  const sources = data.sources
+  if (sources === undefined) {
+    if (rule.requireSources) {
+      failures.push(`${rel}: missing frontmatter \`sources\` (code paths this doc describes)`)
+    }
+  } else if (!Array.isArray(sources) || sources.length === 0 || sources.some((s) => typeof s !== 'string')) {
```
Comment (kangfenmao) [inline scripts/verify-doc-frontmatter.ts:37]: > This review comment was translated automatically.

**[A2]** Here only non-string elements are excluded, `sources: [""]` would still enter the loop below: an empty string is neither an absolute path nor contains `..`, and `path.join(repoRoot, "")` points to an existing repository root directory, so invalid `sources` would pass the gate. This breaks the contract that each reference must be associated with a real code path, and subsequent reverse lookups based on `sources` cannot trust this field. Please also reject `source.trim() === ""`, and add a regression test for empty string entries.

---
<details>
<summary>Original Content</summary>

**[A2]** 这里仅排除了非字符串元素，`sources: [""]` 仍会进入下面的循环：空字符串既不是绝对路径也不含 `..`，而 `path.join(repoRoot, "")` 指向存在的仓库根目录，因此无效的 `sources` 会通过 gate。这样每份 reference 必须关联真实代码路径的契约就无法成立，后续基于 `sources` 的反向查找也不能信任该字段。请同时拒绝 `source.trim() === ""`，并补一个空字符串条目的回归测试。
</details>

**Review by kangfenmao — CHANGES_REQUESTED** [review]
> This review was translated automatically.

The frontmatter gate in the current head still allows empty string source to bypass validation; please fix this edge case and add regression tests before merging.

---
<details>
<summary>Original Content</summary>

当前 head 的 frontmatter gate 仍允许空字符串 source 绕过校验；请修复该边界条件并补充回归测试后再合并。
</details>

**Review by kangfenmao — APPROVED** [review]
当前 head 已复核，空 source 的边界条件已正确修复并补充回归测试，未发现其他需要修改的问题。
